[3단계 - 로또] 이지윤 미션 제출합니다.

dd-jiyun opened the pull request:
안녕하세요, 주노 🔥

2단계에서 받은 강력한 리뷰들 덕분에 구조적인 개선을 많이 할 수 있었고, 3단계 구현을 진행하면서 “리팩토링하길 정말 잘했다!” 는 걸 뼈저리게 느꼈습니다...!

이번 단계에서는 기능 추가보다는 기반 구조를 얼마나 잘 확장할 수 있을지, 그리고 의도를 잘 드러내는 설계가 무엇인지 고민하며 진행했습니다.

리팩토링하면서 파일 변경이 많아 리뷰하시기에 힘드실 것 같아 죄송스럽습니다,, 잘 부탁드립니다! 🙇‍♀️

---
## 주요 변경사항
- `PurchaseAmount, Prize` → 단일 `Money`로 통합
- domain 하위 패키지를 관심사에 따라 세분화
- `Rank.SECOND` 처리 로직에서 보너스볼 일치 여부 메시지 분기 처리 추가

## 고민했던 부분
[2단계 리뷰](https://github.com/next-step/java-lotto-clean-playground/pull/111#discussion_r2098382636)를 참고하여 Money라는 VO로 금액 관련 처리를 전담하게 했어요.

금액 연산은 모두 Money 내부에서 처리하도록 해서 도메인의 책임을 좀 더 분명하게 만들려고 했습니다!

아직 VO의 개념을 온전히 이해한 건지는 확신이 없지만, 금액의 불변성, 의미 부여, 일관된 처리라는 관점에서 VO처럼 잘 사용하고 있는지 궁금합니다 😂

## 궁금한 부분
1. Money를 VO로 설계했다고 판단했는데, VO로서 적절하게 쓰고 있는지 확인받고 싶어요!
2. 도메인 하위 패키지 분리는 과한 분해는 아니었는지 궁금합니다. 혹시 더 좋은 기준이 있다면 알고 싶습니다!
3. 메서드 네이밍이나 변수명을 명확하게 하기위해 신경을 썼어요. 혹시 명확하지 않아보이는 곳이 있다면 피드백 부탁드립니다!

dd-jiyun committed:
remove gitkeep file

dd-jiyun committed:
refactor: 금액과 관련된 도메인 하나로 병합 및 패키지 구조 변경

- 기존 상금과 구매 금액을 돈이라는 객체 하나로 관리하도록 수정하였습니다.
- 도메인이 많아질 수록 클래스를 찾는 것이 어려워 패키지 구조를 변경하였습니다.

dd-jiyun committed:
docs: 구현 기능 목록 수정

dd-jiyun committed:
feat: 2등 보너스 볼 추가

Choi-JJunho commented on src/main/java/Application.java:
```diff
@@ -1,31 +1,31 @@
-        int amount = InputView.inputPurchaseAmount();
-        PurchaseAmount purchaseAmount = new PurchaseAmount(amount);
+        String purchaseAmountInput = InputView.inputPurchaseAmount();
+        Money money = new Money(purchaseAmountInput);
```
InputView에서 바로 Money를 반환하는것은 어떻게 생각하시나요?

dd-jiyun replied:
Money는 값 객체이긴 하지만, 결국 도메인에 속한 객체이기 때문에 View에서 직접 다루는 건 조심할 필요가 있다고 생각해요.

만약, Money의 생성 방식이나 내부 구현이 바뀌게 되면 View도 함께 수정해야 하는 상황이 생기게 돼요. 
이렇게 되면 View와 도메인 사이에 불필요한 결합이 생기기 때문에 View에서는 단순 문자열을 입력받아서 넘기는게 안전하다고 생각합니다!

이 문제를 해결하려면 DTO를 이용해볼 수 있을 것 같아요. 주노는 어떻게 생각하세요? 😮

dd-jiyun replied:
엇 근데 지금 구조는 InputView에서 바로 Money를 반환하는 게 아닌데 제가 잘못 이해하고 있는 걸까요?!

Choi-JJunho replied:
만약 그렇게 구조화한다면? 이라는 의미여서 잘 답변해주셨습니다 👍
말씀해주신대로 도메인과 뷰의 결합이 생겨 다소 변경에 용이하지않은 구조가 될거같군요!
Dto로 구분 좋은것같아요👍

Choi-JJunho commented on src/main/java/domain/lotto/Lotto.java:
```diff
@@ -23,26 +24,26 @@ public List<LottoNumber> getNumbers() {
                 .count();
     }
+
+    public boolean contains(final LottoNumber number) {
+        return numbers.contains(number);
+    }
```
래핑을 잘해주셨군요 👍

jisoo78 commented on src/main/java/domain/lotto/Lotto.java:
```diff
@@ -23,26 +24,26 @@ public List<LottoNumber> getNumbers() {
     private void validateLottoCount(final List<LottoNumber> numbers) {
         if (numbers.size() != LOTTO_NUMBER_COUNT) {
-            throw new IllegalArgumentException("로또 번호는 6개여야 합니다.");
+            throw new IllegalArgumentException("로또 번호는 %s개여야 합니다.".formatted(LOTTO_NUMBER_COUNT));
```
%s 를 통해 입력받은 값(0개 ~ 5개)을 예외 처리를 해 깔끔한 코드가 만들어졌네요!

Choi-JJunho commented on src/main/java/domain/lotto/LottoConstant.java:
```diff
@@ -1,6 +1,6 @@
+package domain.lotto;
 
-public class LottoConstant {
+public final class LottoConstant {
```
final class로 선언해주셨군요!

class를 final로 선언하면 어떤 효과가 있을까요?

dd-jiyun replied:
LottoConstant는 외부에서 생성할 수 없도록 private 생성자를 사용했지만, 여기에 final까지 붙이면 클래스를 상속하는 것도 확실히 막을 수 있어요.

유틸성 클래스는 인스턴스를 만들거나 상속해서 쓸 일이 없어서 final로 선언해서 의도를 명확하게 드러내고, 불필요한 확장을 방지하는 효과가 있어요!

Choi-JJunho commented on src/main/java/domain/money/Money.java:
```diff
@@ -0,0 +1,82 @@
+
+public class Money {
+
+    private static final int DECIMAL_PLACES = 2;
```
해당 값을 정의하게된 계기가 궁금합니다!

dd-jiyun replied:
나눗셈 연산을 할 때 BigDecimal의 divide 메서드를 사용하는데, 
나누는 값에 따라 소수점이 무한히 늘어날 수 있어서 `scale`이나 `RoundingMode`를 명시하지 않으면 예외가 발생하더라고요!

그래서 저는 `소수점 아래 몇 자리까지 계산할지`를 명확히 표현해주기 위해 DECIMAL_PLACES라는 상수를 따로 정의했어요.

혹시 네이밍이 명확하지 않아서 어떤 값인지 알기 어려운가요?

Choi-JJunho replied:
아뇨! 충분히 근거있는 선택이라고 생각합니다 :) 
다만 말씀해주신 내용을 테스트로 작성해보면 이해가 더 쉬울것 같아요!
테스트르는 문서의 역할도하니까요 😊

jisoo78 commented on src/main/java/domain/rank/Rank.java:
```diff
@@ -0,0 +1,50 @@
+        if (matchCount == 5 && bonusMatch) {
+            return SECOND;
+        }
+        if (matchCount == 5) {
```
early return을 사용해 들여쓰기를 줄이고 가독성을 높였네요!

Choi-JJunho commented on src/main/java/domain/money/Money.java:
```diff
@@ -0,0 +1,82 @@
+
+    public Money add(final Money other) {
+        return new Money(this.amount.add(other.amount).toPlainString());
+    }
```
Money는 무조건 String만 받아야한다면 모든 자료형을 String으로 변형해줘야하는 번거로움이 생길 것 같아요 🤔
정수 혹은 문자열 혹은 BigDecimal을 생성자로 받아도 Money로 표현할 수 있게 만들어주려면 어떻게 해볼 수 있을까요?

> 키워드: 오버로딩

dd-jiyun replied:
금액(로또 상금, 구매 금액)과 관련된 VO라서 문자열로 입력을 받아 BigDecimal로 변환하는 방식이 가장 안전하고 명확한 처리라고 생각했어요.

주노의 말씀처럼 들어오는 모든 자료형을 String으로 변환해야하는 번거로움이 있는 건 저도 공감합니다!
하지만 금액 입력은 대부분은 숫자 형식으로 들어오게 될 텐데, 다른 타입까지 고려해서 오버로딩하는 게 좋을까요?

Choi-JJunho replied:
저는 자주 등장하는 int, long, String정도에 대해서는 필요할 수도 있다고 생각해요!
경우에 따라 오버로딩을하거나 정적팩토리메소드로 표현하는 방식도 생각해볼 수 있을것같네요 🤔

dd-jiyun replied:
int나 long 타입에 대해 오버로딩을 고려한 특별한 상황이나 이유가 있을까요?
어떤 케이스에서 이런 오버로딩이 필요하다고 판단하셨는지 주노의 생각이 궁금합니다!

Choi-JJunho replied:
금액이 들어올때 숫자형식으로 들어올 수 있다는 생각이였어요!
하지만 상황에따라 필요없다면 굳이 만들지 않아도되는 영역이라 상황에맞게 선택하는 부분이 될것같네요 ㅎㅎ

Choi-JJunho commented on src/main/java/domain/money/Money.java:
```diff
@@ -0,0 +1,82 @@
+        if (isZeroOrNegative) {
+            throw new IllegalArgumentException("분모는 0보다 커야 합니다.");
+        }
+    }
```
일반적으로 0으로 나누었을 때 어떤 현상이 일어나기를 기대할까요?

외부에서 나누기를 수행하다보면 어떤 상황에서는 0으로 나누는 상황이 일어날 수 있다고 생각이 들어요
이 때 마다 매번 예외를 발생시킨다면 반복적인 예외처리가 일어날 수 있다고 생각하는데요!

제가 생각했을때는 [놀람 최소화 원칙](https://ko.wikipedia.org/wiki/%EB%86%80%EB%9E%8C_%EC%B5%9C%EC%86%8C%ED%99%94_%EC%9B%90%EC%B9%99)을 생각해봤을 때 0을 반환해주는것도 나쁘지 않다고 생각이 들어요

지윤님은 어떻게 생각하시나요?

어디까지나 제 개인적인 생각이니 마음껏 반박해주셔도 좋습니다 😄

dd-jiyun replied:
<img width="195" alt="image" src="https://github.com/user-attachments/assets/b451fe93-f91c-43cf-8bc3-d14035ace610" />

주노가 실제로 보여주신 것처럼 계산기에서 "정의할 수 없음"이라고 표현하거나 0을 반환하는 것이 사용자 입장에서는 당황스럽지 않을 것 같아요.

나눌 때 검증하는 것보다 입력 받는 시점에 검증하는 것이 더 자연스러울 것 같다는 생각도 드네요!

Choi-JJunho commented on src/main/java/domain/money/Money.java:
```diff
@@ -0,0 +1,82 @@
+
+    public static Money zero() {
+        return new Money("0");
+    }
```
정적 메소드로 구현해주셨군요!
하지만 이렇게 구현하면 zero를 호출할 때 마다 매번 Money 인스턴스가 생성될 것 같네요

아래에서 사용해주신 `BigDecimal.ZERO`는 어떻게 구현되어있는지 확인해보고 개선점을 찾아볼까요?

dd-jiyun replied:
말씀 듣고 다시 보니, zero() 호출 시마다 새로운 Money 인스턴스를 생성하는 부분이 불필요한 객체 생성을 유발할 수 있겠다는 생각이 들었어요.

BigDecimal.ZERO의 구현에서는 ` public static final BigDecimal ZERO = new BigDecimal(BigInteger.ZERO, 0, 0, 1);` 처럼 미리 만들어진 불변 인스턴스를 캐싱해서 재사용하고 있더라구요.

`Money`도 static final 상수로 ZERO 값을 캐싱해두면, zero() 호출 시 매번 객체를 새로 만들지 않고 동일한 인스턴스를 재사용할 수 있어서 효율적일 것 같아요!

Choi-JJunho commented on src/main/java/domain/rank/Rank.java:
```diff
@@ -0,0 +1,50 @@
+        this.prize = prize;
+    }
+
+    public static Rank from(final int matchCount, final boolean bonusMatch) {
```
지난번에 이야기나눈 정적팩토리 메소드에 대해 다시한번 확인해보면 좋을 것 같아요 :)

- https://github.com/next-step/java-lotto-clean-playground/pull/111#discussion_r2095150096

dd-jiyun replied:
헉 급하게 리팩토링하다가 놓쳤네요! 수정하겠습니다 👍🏻

Choi-JJunho commented on src/main/java/domain/rank/WinningLotto.java:
```diff
@@ -0,0 +1,60 @@
+    private final LottoNumber bonusNumber;
+
+    public WinningLotto(final String winningNumbers, final String bonusNumber) {
+        validateEmpty(winningNumbers);
+        this.winningNumbers = new Lotto(convertToLottoNumbers(winningNumbers));
+        validateEmpty(bonusNumber);
+        this.bonusNumber = LottoNumber.of(Integer.parseInt(bonusNumber));
+        validateDuplicateWinningNumbers();
```
winningNumbers에 값을 할당한 뒤에 bonusNumber검증을 수행했을 때 검증이 실패하면 winningNumbers에 값을 할당하는 행위가 의미없는 행위가 될 것 같아요

검증 순서를 어떻게 조절해볼 수 있을까요?

dd-jiyun replied:
엇 말씀 듣고 보니 정말 그렇네요!
검증이 끝나기도 전에 객체를 먼저 생성해버리면, 나중에 예외가 발생했을 때 이미 생성된 값은 무의미해질 수 있겠어요 😮

모든 입력값에 대한 검증이 완료된 이후에 필요한 도메인 객체를 생성하도록 순서를 조정하는 게 더 안정적인 흐름인 것 같아요!

Choi-JJunho commented on src/main/java/domain/lotto/Lotto.java:
```diff
@@ -1,6 +1,6 @@
-package domain;
+package domain.lotto;
```
> 도메인 하위 패키지 분리는 과한 분해는 아니었는지 궁금합니다. 혹시 더 좋은 기준이 있다면 알고 싶습니다!

역할별로 잘 나눠주셨다고 생각합니다 👏👏👏

Choi-JJunho commented on src/main/java/domain/money/Money.java:
```diff
@@ -0,0 +1,82 @@
+import java.math.RoundingMode;
+import java.util.Objects;
+
+public class Money {
```
> Money를 VO로 설계했다고 판단했는데, VO로서 적절하게 쓰고 있는지 확인받고 싶어요!

값을 표현하고, 불변을 유지하고있는 점에서 VO로서 잘 활용하고 있다고 생각합니다 👍 

한가지 첨언하자면 VO라면 record로 생성해봐도 좋은 시도일 것 같아요!
toString 등을 굳이 재정의하지 않아도 된다는 점을 가져가볼만 하다고 생각합니다

dd-jiyun replied:
오! record로 생성해도 좋을 것 같아요 감사합니다 😆

Choi-JJunho reviewed (commented):
고생하셨습니다~!
지난번에 말씀드렸던 Money로 감싸보는 시도가 인상적이네요 👍
몇가지 코멘트 남겼으니 확인 부탁드립니다!

izzy80 commented:
3단계 수고 많으셨어요 😊 지윤님의 고민이 잘 녹아든 설계 같아요! 남은 단계도 화이팅입니다 💪

dd-jiyun committed:
refactor: VO 객체 record로 변경

- hashCode나 toString를 재정의하지 않고 사용할 수 있는 장점을 위해 변경하였습니다.
- 불필요한 인스턴스 생성을 방지하기 위해 정적 상수를 캐싱하였습니다.

dd-jiyun committed:
refactor: 정적 팩토리 메서드 네이밍 변경

- 정적 팩토리 메서드의 네이밍 컨벤션에 맞게 변경하였습니다.

dd-jiyun committed:
refactor: 당첨 번호 및 보너스 번호 검증 로직 순서 변경

dd-jiyun reviewed (commented):
리뷰 답변 늦어서 죄송합니다! 
꼼꼼한 리뷰 항상 감사해요 🙇🏻‍♀️

dd-jiyun committed:
refactor: 당첨 번호 파싱 객체 생성

- 문자열에 의존적이었던 당첨 번호 파싱을 분리하였습니다.

dd-jiyun committed:
refactor: final 키워드 추가

dd-jiyun committed:
refactor: 0으로 나눌 시 예외가 아닌 0으로 반환

- 놀람 최소화 원칙에 따라 예외보다는 0을 반환하는 것으로 변경하였습니다.

dd-jiyun committed:
refactor: 당첨 통계 메서드 분리

- 각 메서드가 한 가지 역할에만 집중할 수 있도록 분리했습니다.

dd-jiyun committed:
refactor: BigDecimal setScale 테스트 코드 작성
